### ISSUE TITLE
[BUG] scipy.special.gammaln : Output does not match `math.lgamma`

### ISSUE THREAD

**Issue by kshitij12345** [scipy/scipy#13399]
In notes, the documentation says
> It is the same function as the Python standard library function math.lgamma.

Reference: https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.gammaln.html

However this is not true when input is `-inf`.
```python
>>> import math
>>> import scipy.special
>>> scipy.__version__
'1.5.4'
>>> math.lgamma(float('-inf'))
inf
>>> scipy.special.gammaln(float('-inf'))
-inf
```

**Comment by WarrenWeckesser** [comment]
Shouldn't log(gamma(-inf)) be nan?  The limit as x -> -inf of gamma(x) does not exist, because at each negative integer, the function blows up.  I don't see how one can justify either inf or -inf as the value of log(gamma(-inf)).

**Comment by kshitij12345** [comment]
@WarrenWeckesser Thanks for the response.

Looking at the Wolfram plot for Log Gamma (note : without the abs)
![Screenshot from 2021-02-20 13-29-14](https://user-images.githubusercontent.com/19503980/108588529-b9ca0f00-737f-11eb-906c-1a8f8cd2c108.png)

It looks like for negative integers it blows up to `+inf`.

Besides python's `math`, C++ standard agrees with the same.

Reference: 
https://en.cppreference.com/w/cpp/numeric/math/lgamma
https://mathworld.wolfram.com/LogGammaFunction.html

**Comment by WarrenWeckesser** [comment]
> It looks like for negative integers it blows up to +inf.

Yes, as x approaches a negative integer, abs(gamma(x)) blows up to +inf, and therefore so does log(abs(gamma(x))).  But between those integers, gamma(x) is finite, and so is log(gamma(x)).  So I don't see a mathematical justification for choosing `log(gamma(x))` to be +inf.  (Maybe because any 64 bit floating point number with sufficiently large magnitude is an integer?  Seems like a stretch.)

Having said that, it does appear that returning +inf is the accepted convention:

* The web page https://en.cppreference.com/w/cpp/numeric/math/lgamma says about `lgamma`:  "if the argument is ±∞, +∞ is returned." 
* `lgamma` is specified as part of the C99 math library, but I wasn't able to find an explicit statement of how the argument -inf should be handled.  The N1256 draft says "The lgamma functions compute the natural logarithm of the absolute value of gamma of x. A range error occurs if x is too large. A range error may occur if x is a negative integer or zero." and N1570 (for C11) says "The lgamma functions compute the natural logarithm of the absolute value of gamma of x. A range error occurs if x is too large. A pole error may occur if x is a negative integer or zero."  However, a test program shows that gcc's `lgamma` does, in fact, return `+inf` when the argument is `-inf`.
* As noted earlier, in Python, `math.lgamma(-math.inf)` returns inf.

**Comment by WarrenWeckesser** [comment]
`scipy.special.gamma(-np.inf)` also returns -inf, and that is also likely not correct.  In this case, it looks like `nan` is the expected result:
* At https://en.cppreference.com/w/cpp/numeric/math/tgamma, we have "If the argument is -∞, NaN is returned and FE_INVALID is raised."
* In gcc's math library, `tgamma` returns `nan` when the argument is -inf.
* In Python, `math.gamma(-math.inf)` raises `ValueError: math domain error`.

**Comment by mdickinson** [comment]
[Apologies for the late comment; I just discovered this issue]

@WarrenWeckesser 

> * `lgamma` is specified as part of the C99 math library, but I wasn't able to find an explicit statement of how the argument -inf should be handled.

There's an explicit statement in annex F: C99 §F.9.5.3 says:

> lgamma(−∞) returns +∞.

And that's why I implemented it this way for Python's `math.lgamma`. (So Python's behaviour shouldn't be seen as providing an independent data point here: Python's choice was directly based on the C standard.)

@WarrenWeckesser 

> So I don't see a mathematical justification for choosing `log(gamma(x))` to be +inf [...]

I agree that it's a little odd. My guess is that the "every sufficiently large float is an integer" justification you mention _is_ the reason, but it seems questionable, and I imagine that if IEEE 754 were to standardise `lgamma`, they'd make a different choice.

For comparison, IEEE 754 says that `sinPi` and `cosPi` should signal "invalid operation" for infinite inputs, even though the "large floats are (even) integers" could have been applied there too. On the other hand, IEEE 754 considers `pow(-1, -∞)` to be valid (giving `1`), even though `pow(-1, x)` is invalid for all non-integer negative `x`.

**Comment by mdhaber** [comment]
@steppi I think the action here is to make `gamma` and `gammaln` consistent and document it. Besides the results given above, the result of evaluating the gamma function at -infinity in Mathematical and R is also NaN. (Wolfram Alpha says the limit is 0, but I can't reproduce that with Mathematica itself, and given that the value oscillates I don't see how that would be right.)

**Comment by steppi** [comment]
What an odd situation.  I found the following passage in Jean-Michel Muller's *Handbook of Floating-Point Arithmetic*.

*In its conditionally normative Annex F, the C standard also specifies the val-
ues of the elementary functions for particular arguments. These definitions,
which came from C99 (thus before IEEE 754-2008 came out with its own rec-
ommendations) and are explained in the C99 rationale [275], can cause some
head scratching. Consider for instance, the following requirement for the pow
function: pow(−1, ±∞) = 1. This may be a bit puzzling until one understands
that any sufficiently large binary floating-point number is an even integer:
hence, “by taking the limit,” the prescribed value of 1. The result of pow(x, ±0)
is 1 for any x, including 0, because this result was found to be more useful
in practice than a NaN.8 Moreover, when a partial function is constant, such
as pow(x, ±0), it was chosen to return this constant even when the variable is
a quiet NaN (let us recall that C11 does not specify the behavior of signaling
NaNs): hence, pow(NAN, ±0) = 1.
Concerning the special cases, IEEE 754-2008 chose to remain compatible
with C99, but added three variants of pow: powr (whose special cases are
derived by considering that it is defined as ey log x ) and pown/rootn (which
deal with integral exponents only).*

So it seems that the C99 standard decided on a principle that when a partial function $f(x | \mathrm(args))$ is constant when $x$ ranges over finite floating point values, then it remains constant when $x$ takes on exceptional values, including even a `quietNaN`. Note that the rationale above specifically mentions that every sufficiently large in magnitude float is an even integer. IEEE-754 chose to remain consistent with the C99 standard. `sinpi` isn't in the C99 standard, so IEEE-754 was free to go with the behavior they thought was best. So I agree with @mdickinson that if IEEE-754 were to standardize `lgamma`, they would have most likely gone with a different behavior, but only in a just world where they standardized it first.

I support disregarding C99's principle regarding constant partial functions in `special` (and other similar recommendations which don't make sense from a strictly mathematical standpoint if we find any) and just adding notes in the documentation when this is the case, as @mdhaber suggested.
